src/ring/adapter/undertow/middleware/session.clj

```diff
@@ -71,20 +71,34 @@
   value stored in the `io.undertow.server.session.Session` from the
   associated handler"
   [handler options]
-  (let [fallback (delay (ring-session/wrap-session handler options))]
-    (fn [request]
-      (if-let [^HttpServerExchange exchange (:server-exchange request)]
-        (let [response (handler
+  (let [fallback (delay (ring-session/wrap-session handler options))
+        side-effects! (fn [response ^HttpServerExchange exchange]
+                        (when (contains? response :session)
+                          (if-let [data (:session response)]
+                            (when-let [session (get-or-create-session exchange)]
+                              (set-ring-session! session data))
+                            (when-let [session (Sessions/getSession exchange)]
+                              (.invalidate session exchange)))))]
```

Comment:
Let's pull this out into a separate function since it can be initialized separately from the middleware.

Reply:
There you are